src/Microsoft.VisualStudio.Threading.Shared/AsyncBarrier.cs

```diff
@@ -52,21 +50,33 @@ public AsyncBarrier(int participants)
         /// <returns>An awaitable.</returns>
         public Task SignalAndWait()
         {
-            var tcs = new TaskCompletionSource<bool>();
-            this.waiters.Push(tcs);
-            if (Interlocked.Decrement(ref this.remainingParticipants) == 0)
+            lock (this.waiters)
```

Comment:
The use of locks in code we know is executing concurrently has been troubling me recently (it keeps showing up much higher than I would hope for in PerfView). It's fairly straightforward to rewrite this method as lock-free (interlocked update) as long as you allow allocations. For the fast path, the easy implementation would allocate one of the following per call to `SignalAndWait`:

```csharp
class Node {
  int count;
  TaskCompletionSource<EmptyStruct> current;
  Node next;
}
```